typing/typecore.ml

```diff
@@ -4660,7 +4654,7 @@ and type_cases ?in_function env ty_arg ty_res partial_flag loc caselist =
   let in_function = if List.length caselist = 1 then in_function else None in
   let cases =
     List.map2
-      (fun (pat, _, (ext_env, unpacks)) {pc_lhs; pc_guard; pc_rhs} ->
+      (fun (pat, _, (ext_env, unpacks)) {pc_lhs = _; pc_guard; pc_rhs} ->
```

Comment:
Why this change? Is there anything new here?

Reply:
Because `contains_gadt` is now called on `pat` and not `pc_lhs` which is not used anymore.

Reply:
It took me a while to remember that `pc_lhs` (in this context) is from the pasetree, and `pat` is from the typed tree...
OK, now I understand the change.

Note that it also means that you could delay building the copied environment to after typing the patterns (since their typing does not depend on values in the environment), and only copy identifiers used in GADT branches.
If this copying is the reason for the slowdown, this would completely remove it.
This could be in particular useful for let-expressions matching on a single-case sum-type.

Reply:
> It took me a while to remember that `pc_lhs` (in this context) is from the pasetree, and `pat` is from the typed tree...

Sorry, should have made that clearer.

> Note that it also means that you could delay building the copied environment to after typing the patterns (since their typing does not depend on values in the environment), and only copy identifiers used in GADT branches.
If this copying is the reason for the slowdown, this would completely remove it.

Yes, I thought about that. But it's not completely clear whether it would be a win or not. At the moment we duplicate all the identifiers appearing in patterns containing a constructor, but we do it only once. Whereas if we did it after having typed the pattern, we would do it only for branches actually containing GADT constructors, but we might have to copy the same identifiers over and over again (because each branch gets its own environment).
It's probably worth a try though; but I'll do it separately from this PR if that's alright.

Reply:
No, you could collect the identifiers from all branches that actually contained gadt patterns, and copy them only once. I.e. basically the same as now, but move it just after the type-checking of patterns (but before calling the pattern_force). There is no need to copy the environment before.  Could you give it a try?

Reply:
I think I see what you mean. I'll give it a try sometime in the next few days and report back.